docs/EVALUATION.md

```diff
@@ -0,0 +1,249 @@
+Experiment: training time vs. training batch size
+
+# RFC: How and when to run evaluations
+
+## Motivation
+
+I) We want to be able to determine how well our models are performing according to some metrics like accuracy, precision, recall, F1-score, etc.
+
+II) We also want to be able to determine how well our models are performing over time. This will allow us to determine if our models are improving, degrading, or staying the same over time.
+
+III) We are facing two use cases that require us to run evaluations:
+
+1) We want to run evaluations in an online fashion so we can use those evaluations for triggering policies. For example, we may want to trigger a policy that will retrain a model if the model's performance degrades below a certain threshold.
+
+2) We want to run evaluations in an offline fashion for experiments so we determine the performance of a model **[1] at a certain point in time** as well as the **[2] total pipeline performance** aggregated over the whole time of the pipeline run (= pipeline performance).
+
+## Setting
+
+We are looking at a completed pipeline run that produced multiple triggers and therefore multiple models. We want to evaluate the performance of each model.
+
+```mermaid
+gantt
+    title A Gantt Diagram
+    dateFormat YYYY-MM-DD
+    axisFormat %Y
+    section models
+        Trigger1 :milestone, 1930, 0d
+        model 1          :a1, 1930, 10y
+        Trigger2 :milestone, :after a1
+        model 2          :a2, after a1, 5y
+        Trigger3 :milestone, :after a2
+        model 3          :a3, after a2, 5y
+        Trigger4 :milestone, :after a3
+        model 4          :a4, after a3, 25y
+        Trigger5 :milestone, :after a4
+        model 5          :a5, after a4, 25y
+        Trigger6 :milestone, :after a5
+        model 6          :a6, after a5, 5y
+        Trigger7 :milestone, :after a6
+        model 7          :a7, after a6, 5y
+```
+
+## Evaluation methodology
+
+An evaluation methodology consists of three main components:
+
+1) The evaluation metric (e.g. F1, accuracy, precision, recall, etc.)
+2) The evaluation data (set of samples out of the evaluation split used for single model evaluation)
+3) The points in time when the evaluations are run
+
+### Use case 1: online evaluation
+
+In this case we have to be extremely cautious to keep evaluation costs low as the are part of the trigger policy evaluation costs (or at least part of the core pipeline that cannot be disregarded like the experiment model evaluations).
+
+During the core pipeline execution (not including experiment mode pipeline steps), we cannot access future data for evaluation. Also we don't have an evaluation split available. Therefore, we need to use the training data up to the current point in time for evaluation. This is not ideal but it is the only way to evaluate the model in an online fashion.
+
```

Comment:
I didn't totally get it. You mean we cannot use evaluation data at all?
I think we can use evaluation data up to the current point (as long as not access future eval data)

Reply:
No, we can definitely run evaluations in that mode. However, in production mode with Modyn, we only use one dataset for training and do not have an additional "testing" or evaluation dataset. When conducting evaluations to provide accuracy feedback into the pipeline (e.g., triggers), we can, as you mentioned, allow the evaluation to use past data. I wanted to point this out to keep it in mind, but for now, I don't intend to focus on this aspect.

Reply:
I see thanks!